mlir/lib/Gradient/IR/GradientOps.cpp

```diff
@@ -43,7 +55,7 @@ LogicalResult verifyGradInputs(OpState *op_state, func::FuncOp callee, ValueRang
                << "expected " << fnType.getNumInputs() << " but got " << fnArgs.size();
 
     for (unsigned i = 0; i < fnArgs.size(); ++i)
-        if (fnArgs[i].getType() != fnType.getInput(i))
+        if (!shapeEqual(fnArgs[i].getType(), fnType.getInput(i)))
```

Comment:
It's the latter, we otherwise run into an issue where we can't lower because `func-bufferize` converts the argument types to memrefs but the backprop op still expects the arguments to be tensors (or vice versa if we run `gradient-bufferize` first).